### PR TITLE
JAMES-2586 Implement SieveQuotaRepository backed by Postgres

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve https://github.com/linagora/james-project/issues/4929

TODO: guice binding for PostgresQuota

### CODE REVIEW

File: server/data/data-postgres/src/test/java/org/apache/james/sieve/postgres/PostgresSieveRepositoryTest.java [hunk 1]
```diff
@@ -17,30 +17,39 @@
...
-        sieveRepository = new JPASieveRepository(JPA_TEST_CLUSTER.getEntityManagerFactory());
+        sieveRepository = new PostgresSieveRepository(JPA_TEST_CLUSTER.getEntityManagerFactory(),
+            new PostgresSieveQuotaDAO(new PostgresQuotaCurrentValueDAO(postgresExtension.getPostgresExecutor()),
+                new PostgresQuotaLimitDAO(postgresExtension.getPostgresExecutor())));
     }
 
     @AfterEach
     void tearDown() {
-        JPA_TEST_CLUSTER.clear("JAMES_SIEVE_SCRIPT", "JAMES_SIEVE_QUOTA");
+        JPA_TEST_CLUSTER.clear("JAMES_SIEVE_SCRIPT");
```
Comment (chibenwa) [inline server/data/data-postgres/src/test/java/org/apache/james/sieve/postgres/PostgresSieveRepositoryTest.java:52]: Not needed, right?

File: server/data/data-postgres/src/test/resources/persistence.xml [hunk 2]
```diff
@@ -27,12 +27,10 @@
        <provider>org.apache.openjpa.persistence.PersistenceProviderImpl</provider>
         <jta-data-source>osgi:service/javax.sql.DataSource/(osgi.jndi.service.name=jdbc/james)</jta-data-source>
         <class>org.apache.james.domainlist.jpa.model.JPADomain</class>
-        <class>org.apache.james.user.jpa.model.JPAUser</class>
         <class>org.apache.james.rrt.jpa.model.JPARecipientRewrite</class>
         <class>org.apache.james.mailrepository.jpa.model.JPAUrl</class>
         <class>org.apache.james.mailrepository.jpa.model.JPAMail</class>
-        <class>org.apache.james.sieve.jpa.model.JPASieveQuota</class>
-        <class>org.apache.james.sieve.jpa.model.JPASieveScript</class>
+        <class>org.apache.james.sieve.postgres.model.JPASieveScript</class>
```
Comment (chibenwa) [inline server/data/data-postgres/src/test/resources/persistence.xml:33]: Remove it too?

File: server/data/data-postgres/src/test/java/org/apache/james/sieve/postgres/PostgresSieveRepositoryTest.java (hunk 1 above)
Comment (Arsnael) [reply server/data/data-postgres/src/test/java/org/apache/james/sieve/postgres/PostgresSieveRepositoryTest.java:52]: Sieve script migration: other task => https://github.com/linagora/james-project/issues/4928

**Comment by quantranhong1999** [comment]
> <openjpa-3.2.0-r6f721f6 nonfatal user error> org.apache.openjpa.persistence.ArgumentException: This configuration disallows runtime optimization, but the following listed types were not enhanced at build time or at class load time with a javaagent: "
org.apache.james.sieve.postgres.model.JPASieveScript".

Before I had some issue running JPA test when I changed some JPA code. Rebuild the code should help the test runs normally, or add the following JVM property to run configuration: `-Dopenjpa.RuntimeUnenhancedClasses=supported`.

But now even after rebuild and on the CI, the JPA tests still can not run normally. Will need investigate on this...

**Comment by Arsnael** [comment]
Perhaps the error is elsewhere? If you look at some other tests => https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1819/1/testReport/junit/org.apache.james.sieve.postgres/PostgresSieveRepositoryTest/haveSpaceShouldNotThrowWhenAttemptToReplaceOtherScript/

```
An error occurred while parsing the query filter "SELECT sieveScript FROM JamesSieveScript sieveScript WHERE sieveScript.username=:username". Error message: The name "JamesSieveScript" is not a recognized entity or identifier. Perhaps you meant JamesMailStore, which is a close match. Known entity names: [JamesMailStore, JamesDomain, JamesMailRepos, JamesRecipientRewrite]
```

**Comment by quantranhong1999** [comment]
> An error occurred while parsing the query filter "SELECT sieveScript FROM JamesSieveScript sieveScript WHERE sieveScript.username=:username". Error message: The name "JamesSieveScript" is not a recognized entity or identifier. Perhaps you meant JamesMailStore, which is a close match. Known entity names: [JamesMailStore, JamesDomain, JamesMailRepos, JamesRecipientRewrite]

Hmm, I guess because somehow `org.apache.james.sieve.postgres.model.JPASieveScript` was not enhanced at build time, therefore OpenJPA does not accept it as an entity as well as create a table for it. Maybe I misconfigured this class somewhere, will try to find it.

File: server/apps/postgres-app/src/main/resources/META-INF/persistence.xml [hunk 3]
```diff
@@ -34,8 +34,7 @@
         <class>org.apache.james.mailrepository.jpa.model.JPAUrl</class>
         <class>org.apache.james.mailrepository.jpa.model.JPAMail</class>
         <class>org.apache.james.rrt.jpa.model.JPARecipientRewrite</class>
-        <class>org.apache.james.sieve.jpa.model.JPASieveQuota</class>
-        <class>org.apache.james.sieve.jpa.model.JPASieveScript</class>
+        <class>org.apache.james.sieve.postgres.model.JPASieveScript</class>
```
Comment (Arsnael) [inline server/apps/postgres-app/src/main/resources/META-INF/persistence.xml:37]: Suspecting that might cause your issue... I see in the code pom.xml that james-server-guice-sieve-jpa module is still imported. Would not be surprised of shenanigans here. 

Try to switch this back to jpa package and test again to see maybe?

**Review by vttranlina — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by quantranhong1999** [comment]
TODO: guice binding for PostgresSieveQuota

File: server/data/data-postgres/pom.xml [hunk 4]
```diff
@@ -155,9 +155,7 @@
                 <artifactId>openjpa-maven-plugin</artifactId>
                 <version>${apache.openjpa.version}</version>
                 <configuration>
-                    <includes>org/apache/james/sieve/jpa/model/JPASieveQuota.class,
-                        org/apache/james/sieve/jpa/model/JPASieveScript.class,
-                        org/apache/james/user/jpa/model/JPAUser.class,
+                    <includes>org/apache/james/sieve/postgres/model/JPASieveScript.class,
```
Comment (chibenwa) [inline server/data/data-postgres/pom.xml:158]: At the very least I want to see a Github issue for removing that JPA entity...

Comment (quantranhong1999) [reply server/data/data-postgres/pom.xml:158]: cf https://github.com/linagora/james-project/issues/4928#issuecomment-1829281340

**Comment by quantranhong1999** [comment]
squashed.

**Review by chibenwa — APPROVED** [review]

**Pull request merged**